test/new-e2e/utils/e2e/e2e.go

```diff
@@ -0,0 +1,225 @@
+// See [Suite] for an example of the usage.
+package e2e
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"reflect"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-agent/test/new-e2e/utils/clients"
+	"github.com/DataDog/datadog-agent/test/new-e2e/utils/credentials"
+	"github.com/DataDog/datadog-agent/test/new-e2e/utils/infra"
+	"github.com/DataDog/test-infra-definitions/aws"
+	"github.com/DataDog/test-infra-definitions/common/config"
+	"github.com/DataDog/test-infra-definitions/common/vm"
+	"github.com/pulumi/pulumi/sdk/v3/go/auto"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"golang.org/x/crypto/ssh"
+)
+
+// Suite manages the environment creation and runs E2E tests.
+// It is implemented as a [testify Suite] and each test defines the environment and the E2E logic.
+// Example of usage:
+//
+//	type vmSuite struct {
+//		e2e.Suite
+//	}
+//
+//	func TestMyEnv(t *testing.T) {
+//		suite.Run(t, &vmSuite{Suite: e2e.NewSuite("aws/sandbox", "my-test")})
+//	}
+//
+//	func (v *vmSuite) Test1() {
+//		v.Run(
+//			func(ctx *pulumi.Context) (*vm.VM, error) {
+//				return ec2vm.NewEc2VM(ctx) // The environment to create. An Empty VM in this example.
+//			},
+//			func(client *e2e.Client) { // The E2E test.
+//				output := client.Execute("ls /tmp")
+//				require.NotEmpty(v.T(), output)
+//			},
+//		)
+//	}
+//
+// Suite leverages pulumi features to compute the differences between the previous
+// environment and the new one to make environment updates faster.
+//
+// [testify Suite]: https://pkg.go.dev/github.com/stretchr/testify/suite
+type Suite struct {
+	envName       string
+	stackName     string
+	ctx           context.Context
+	destroyEnv    bool
+	fullStackName string
+
+	// These fields are initialized in SetupTest
+	suite.Suite
+	configMap auto.ConfigMap
+	require   *require.Assertions
+	sshKey    string
+}
+
+// KeepEnv prevents [Suite] from destroying the environment at the end of the test suite.
+// When using this option, you have to manually destroy the environment.
+func KeepEnv() func(*Suite) {
+	return func(p *Suite) { p.destroyEnv = false }
+}
+
+// NewSuite creates a new [Suite].
+// envName can be either "aws/sandbox" or "az/sandbox".
+// stackName is the name of the stack and should be unique.
+// options are optional parameters for example [e2e.KeepEnv].
+func NewSuite(envName string, stackName string, options ...func(*Suite)) Suite {
+	testSuite := Suite{
+		envName:    envName,
+		stackName:  stackName,
+		ctx:        context.Background(),
+		configMap:  auto.ConfigMap{},
+		destroyEnv: true,
+	}
+	for _, o := range options {
+		o(&testSuite)
+	}
+	return testSuite
+}
+
+// SetupTest method is run before each test in the suite.
+// This function is called by [testify Suite].
+//
+// [testify Suite]: https://pkg.go.dev/github.com/stretchr/testify/suite
+func (suite *Suite) SetupTest() {
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**❓ question**
Why do you need to run this setup before each test in the same suite ? Do these parameters change ?

Reply:
Comment added.